File: frontend/src/services/websocketService.js
// Service WebSocket Client pour CityFlow (Web React)

class CityFlowWebSocketService {
  constructor() {
    this.ws = null;
    this.listeners = new Map();
    this.status = "disconnected"; // 'connected' | 'connecting' | 'disconnected'
    this.statusListeners = new Set();
    this.reconnectTimeout = null;
    this.subscribedCity = null;
  }

  connect(url = "ws://localhost:3000/ws") {
    if (this.ws && (this.ws.readyState === WebSocket.OPEN || this.ws.readyState === WebSocket.CONNECTING)) {
      return;
    }

    this._setStatus("connecting");

    try {
      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        console.log("⚡ [CityFlow WebSocket] Connecté avec succès au serveur temps réel.");
        this._setStatus("connected");
        if (this.subscribedCity) {
          this.subscribeCity(this.subscribedCity);
        }
      };

      this.ws.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);
          this._emit(data.type, data);
          this._emit("*", data); // Écouteur global
        } catch (err) {
          console.error("❌ [CityFlow WebSocket] Erreur décodage message", err);
        }
      };

      this.ws.onclose = () => {
        this._setStatus("disconnected");
        this._scheduleReconnect(url);
      };

      this.ws.onerror = (err) => {
        console.warn("⚠️ [CityFlow WebSocket] Erreur connexion", err);
        this._setStatus("disconnected");
      };
    } catch (err) {
      this._setStatus("disconnected");
      this._scheduleReconnect(url);
    }
  }

  _scheduleReconnect(url) {
    if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
    this.reconnectTimeout = setTimeout(() => {
      console.log("🔄 [CityFlow WebSocket] Tentative de reconnexion automatique...");
      this.connect(url);
    }, 3000);
  }

  _setStatus(newStatus) {
    this.status = newStatus;
    this.statusListeners.forEach((fn) => fn(newStatus));
  }

  subscribeCity(city) {
    this.subscribedCity = city;
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.send({ type: "SUBSCRIBE_CITY", city });
    }
  }

  send(data) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }

  on(eventType, callback) {
    if (!this.listeners.has(eventType)) {
      this.listeners.set(eventType, new Set());
    }
    this.listeners.get(eventType).add(callback);
    return () => this.off(eventType, callback);
  }

  off(eventType, callback) {
    if (this.listeners.has(eventType)) {
      this.listeners.get(eventType).delete(callback);
    }
  }

  onStatusChange(callback) {
    this.statusListeners.add(callback);
    callback(this.status);
    return () => this.statusListeners.delete(callback);
  }

  _emit(eventType, data) {
    if (this.listeners.has(eventType)) {
      this.listeners.get(eventType).forEach((cb) => {
        try {
          cb(data);
        } catch (e) {
          console.error("Erreur listener callback WS", e);
        }
      });
    }
  }

  disconnect() {
    if (this.reconnectTimeout) clearTimeout(this.reconnectTimeout);
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this._setStatus("disconnected");
  }
}

// Instance Singleton
export const wsService = new CityFlowWebSocketService();
export default wsService;

File: mobile/lib/core/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:latlong2/latlong.dart';

enum WsConnectionStatus { connected, connecting, disconnected }

class CityFlowWebSocketService {
  static final CityFlowWebSocketService _instance = CityFlowWebSocketService._internal();
  factory CityFlowWebSocketService() => _instance;
  CityFlowWebSocketService._internal();

  WebSocket? _socket;
  WsConnectionStatus _status = WsConnectionStatus.disconnected;
  Timer? _reconnectTimer;
  String? _subscribedCity;
  bool _isManualDisconnect = false;

  // Stream controllers
  final _statusController = StreamController<WsConnectionStatus>.broadcast();
  final _messageController = StreamController<Map<String, dynamic>>.broadcast();

  // Getters
  WsConnectionStatus get status => _status;
  Stream<WsConnectionStatus> get statusStream => _statusController.stream;
  Stream<Map<String, dynamic>> get messageStream => _messageController.stream;
  bool get isConnected => _status == WsConnectionStatus.connected;

  static const List<String> _candidateWsUrls = [
    'ws://127.0.0.1:3000/ws',
    'ws://192.168.1.123:3000/ws',
    'ws://10.0.2.2:3000/ws',
    'ws://localhost:3000/ws',
  ];

  static String get defaultWsUrl {
    if (kIsWeb) {
      return 'ws://localhost:3000/ws';
    }
    return _candidateWsUrls.first;
  }

  void connect({String? customUrl}) {
    if (_socket != null &&
        (_status == WsConnectionStatus.connected ||
            _status == WsConnectionStatus.connecting)) {
      return;
    }

    _isManualDisconnect = false;
    _setStatus(WsConnectionStatus.connecting);
    _reconnectTimer?.cancel();

    _tryConnectCandidates(customUrl != null ? [customUrl] : _candidateWsUrls);
  }

  Future<void> _tryConnectCandidates(List<String> candidates) async {
    for (final url in candidates) {
      if (_isManualDisconnect) return;
      try {
        debugPrint('⚡ [CityFlow Mobile WS] Tentative vers $url...');
        _socket = await WebSocket.connect(url).timeout(const Duration(seconds: 2));
        _setStatus(WsConnectionStatus.connected);
        debugPrint('✅ [CityFlow Mobile WS] Connecté au serveur temps réel ($url) !');

        if (_subscribedCity != null) {
          subscribeCity(_subscribedCity!);
        }

        _socket!.listen(
          (data) {
            try {
              if (data is String) {
                final parsed = json.decode(data);
                if (parsed is Map<String, dynamic>) {
                  _messageController.add(parsed);
                }
              }
            } catch (e) {
              debugPrint('❌ [CityFlow Mobile WS] Erreur décodage JSON: $e');
            }
          },
          onDone: () {
            debugPrint('⚠️ [CityFlow Mobile WS] Connexion terminée.');
            _setStatus(WsConnectionStatus.disconnected);
            _scheduleReconnect(url);
          },
          onError: (error) {
            debugPrint('❌ [CityFlow Mobile WS] Erreur socket: $error');
            _setStatus(WsConnectionStatus.disconnected);
            _scheduleReconnect(url);
          },
          cancelOnError: true,
        );
        return;
      } catch (_) {
        // Tenter candidat suivant
      }
    }
    _setStatus(WsConnectionStatus.disconnected);
    _scheduleReconnect(_candidateWsUrls.first);
  }

  void _scheduleReconnect(String url) {
    if (_isManualDisconnect) return;
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(const Duration(seconds: 5), () {
      debugPrint('🔄 [CityFlow Mobile WS] Reconnexion automatique...');
      _tryConnectCandidates(_candidateWsUrls);
    });
  }

  void subscribeCity(String city) {
    _subscribedCity = city;
    send({'type': 'SUBSCRIBE_CITY', 'city': city});
  }

  // ──────────────────────────────────────────────────────────────────────────
  // GESTION DE ROUTE ACTIVE — Notifications géo-contextuelles sur trajet
  // ──────────────────────────────────────────────────────────────────────────

  /// Enregistre la route active du conducteur sur le serveur WebSocket.
  /// Le backend utilisera ces coordonnées pour détecter les incidents
  /// signalés à moins de 200m du trajet et envoyer des alertes ciblées.
  ///
  /// [routePoints] — Liste des coordonnées GPS de la polyligne de navigation
  /// [city]        — Ville de la route (ex: 'Yaoundé')
  /// [driverId]    — Identifiant optionnel du conducteur
  void registerActiveRoute({
    required List<LatLng> routePoints,
    required String city,
    String? driverId,
  }) {
    if (routePoints.length < 2) return;

    final coordinates = routePoints
        .map((p) => [p.latitude, p.longitude])
        .toList();

    send({
      'type': 'REGISTER_ROUTE',
      'routeId': 'route_${DateTime.now().millisecondsSinceEpoch}',
      'coordinates': coordinates,
      'city': city,
      if (driverId != null) 'driverId': driverId,
    });

    debugPrint('🗺️ [WS] Route enregistrée : ${routePoints.length} points ($city)');
  }

  /// Désenregistre la route active (fin de navigation).
  /// Le conducteur ne recevra plus d'alertes d'incident sur trajet.
  void unregisterRoute() {
    send({'type': 'UNREGISTER_ROUTE'});
    debugPrint('🛑 [WS] Route désenregistrée — alertes trajet désactivées');
  }

  // ──────────────────────────────────────────────────────────────────────────

  void send(Map<String, dynamic> data) {
    if (_socket != null && _status == WsConnectionStatus.connected) {
      try {
        _socket!.add(json.encode(data));
      } catch (e) {
        debugPrint('❌ [CityFlow Mobile WS] Erreur envoi: $e');
      }
    }
  }

  void _setStatus(WsConnectionStatus newStatus) {
    _status = newStatus;
    if (!_statusController.isClosed) {
      _statusController.add(newStatus);
    }
  }

  void disconnect() {
    _isManualDisconnect = true;
    _reconnectTimer?.cancel();
    _socket?.close();
    _socket = null;
    _setStatus(WsConnectionStatus.disconnected);
  }

  void dispose() {
    disconnect();
    _statusController.close();
    _messageController.close();
  }
}
